crates/transport/src/vec_codec.rs

```diff
@@ -0,0 +1,154 @@
+use std::{
+    any::TypeId,
+    io::{Cursor, ErrorKind},
+    marker::PhantomData,
+};
+
+use async_trait::async_trait;
+use futures::{AsyncReadExt, AsyncWriteExt};
+use libp2p::request_response;
+use prost::{
+    bytes::{Buf, BytesMut},
+    Message,
+};
+
+pub struct ProtoVecCodec<Req, Res> {
+    _req: PhantomData<Req>,
+    _res: PhantomData<Res>,
+    max_req_size: u64,
+    max_res_size: u64,
+}
+
+impl<Req, Res> ProtoVecCodec<Req, Res> {
+    pub fn new(max_req_size: u64, max_res_size: u64) -> Self {
+        Self {
+            _req: Default::default(),
+            _res: Default::default(),
+            max_req_size,
+            max_res_size,
+        }
+    }
+}
+
+impl<Req, Res> Clone for ProtoVecCodec<Req, Res> {
+    fn clone(&self) -> Self {
+        *self
+    }
+}
+
+impl<Req, Res> Copy for ProtoVecCodec<Req, Res> {}
+
+#[async_trait]
+impl<Req: Message + Default + 'static, Res: Message + Default + 'static> request_response::Codec
+    for ProtoVecCodec<Req, Res>
+{
+    type Protocol = &'static str;
+    type Request = Vec<Req>;
+    type Response = Res;
+
+    async fn read_request<T>(
```

Comment:
If I'm not mistaken, these functions are called while holding the mutable ref to the swarm in the request_response behaviour. Is it okay for the server to process them sequentially? Is the performance good enough?